### ISSUE TITLE
examples communicate-between-panels not working in 8.4

### ISSUE THREAD

**Issue by samant-rags** [gapitio/gapit-htmlgraphics-panel#103]
tried running the example 
https://gapit-htmlgraphics-panel.gapit.io/docs/examples/communicate-between-panels/

but not working

**Comment by ZuperZee** [comment]
Did you add both the panels (dispatcher and receiver) on the same dashboard?

And what version of HTMLGraphics are you using?

**Comment by ZuperZee** [comment]
![communicate-between-panels-importing](https://user-images.githubusercontent.com/39485579/172586308-d8a44ba4-a71d-4673-b38c-5fca3bc7c5d3.gif)

Here is the entire JSON Model
```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "target": {
          "limit": 100,
          "matchAny": false,
          "tags": [],
          "type": "dashboard"
        },
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "links": [],
  "liveNow": false,
  "panels": [
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 8,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 4,
      "options": {
        "SVGBaseFix": true,
        "add100Percentage": true,
        "calcsMutation": "standard",
        "centerAlignContent": true,
        "codeData": "{\n  \"text\": \"Random text\"\n}",
        "css": "* {\n  font-family: Open Sans;\n}\n",
        "dynamicData": false,
        "dynamicFieldDisplayValues": false,
        "dynamicHtmlGraphics": false,
        "dynamicProps": false,
        "html": "<div style=\"text-align: center;\">Recieved value <strong /></div>",
        "onInit": "const valueElt = htmlNode.querySelector(\"strong\");\n\n\nconst doSomething = (e) => {\n  const value = e.detail;\n  valueElt.textContent = value;\n};\n\ndocument.addEventListener(\"htmlgraphics\", doSomething);\n\nhtmlNode.onpanelwillunmount = () => {\n  document.removeEventListener(\"htmlgraphics\", doSomething);\n};\n",
        "onInitOnResize": false,
        "onRender": "",
        "overflow": "visible",
        "panelupdateOnMount": true,
        "reduceOptions": {
          "calcs": [
            "lastNotNull",
            "last",
            "first",
            "firstNotNull",
            "min",
            "max",
            "mean",
            "sum",
            "count",
            "range",
            "delta",
            "step",
            "diff",
            "logmin",
            "allIsZero",
            "allIsNull",
            "diffperc"
          ]
        },
        "renderOnMount": true,
        "rootCSS": ""
      },
      "title": "Panel Title",
      "type": "gapit-htmlgraphics-panel"
    },
    {
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 8
      },
      "id": 2,
      "options": {
        "SVGBaseFix": true,
        "add100Percentage": true,
        "calcsMutation": "standard",
        "centerAlignContent": true,
        "codeData": "{\n  \"text\": \"Random text\"\n}",
        "css": "* {\n  font-family: Open Sans;\n}\n",
        "dynamicData": false,
        "dynamicFieldDisplayValues": false,
        "dynamicHtmlGraphics": false,
        "dynamicProps": false,
        "html": "<button>Dispatch event and pass the metric value <strong /></button>\n",
        "onInit": "const buttonElt = htmlNode.querySelector(\"button\");\nconst valueElt = htmlNode.querySelector(\"strong\");\n\nbuttonElt.onclick = () => {\n  htmlGraphicsEvent = new CustomEvent(\"htmlgraphics\", { detail: valueElt.textContent });\n  document.dispatchEvent(htmlGraphicsEvent);\n};\n",
        "onInitOnResize": false,
        "onRender": "// Sets the value from the first series on every refresh\nconst htmlgraphicsValue = htmlNode.querySelector(\"strong\");\n\nif (htmlgraphicsValue) {\n  const valueField = data.series[0]?.fields[1];\n  if (valueField) {\n    const length = valueField.values.length;\n    htmlgraphicsValue.textContent = valueField.values.get(length - 1);\n  } else {\n    htmlgraphicsValue.textContent = \"No data\"\n  }\n}\n",
        "overflow": "visible",
        "panelupdateOnMount": true,
        "reduceOptions": {
          "calcs": [
            "lastNotNull",
            "last",
            "first",
            "firstNotNull",
            "min",
            "max",
            "mean",
            "sum",
            "count",
            "range",
            "delta",
            "step",
            "diff",
            "logmin",
            "allIsZero",
            "allIsNull",
            "diffperc"
          ]
        },
        "renderOnMount": true,
        "rootCSS": ""
      },
      "title": "Panel Title",
      "type": "gapit-htmlgraphics-panel"
    }
  ],
  "schemaVersion": 35,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "New dashboard",
  "version": 0,
  "weekStart": ""
}
```

**Comment by samant-rags** [comment]
Thanks Zuper- this worked ! 
However, had to changed my Dispatcher panel datasource too.

Also wondering from a design point of view -  if relying on 
dispatchEvent  for multiple fields (say form elements spread over in 2-3 panels as shown in attached image) 
is better or navigating through shadowRoot DOM  Or  is there any other better approach ?


![image](https://user-images.githubusercontent.com/5258222/173170001-f67814ab-5e7b-4b6b-acaa-0ef9dd7ccc87.png)

**Comment by ZuperZee** [comment]
I haven't figured out a way to communicate between panels without using a _global_ variable to attach something to (like document).

With forms you would love to use

```html
<div>
  <shadowroot>
    <form id="my-form">
      <input />
    </form>
  </shadowroot>

  <shadowroot>
    <button form="my-form" type="submit">Submit</button>
  </shadowroot>
</div>
```

But because of shadowroot. The elements can't _see_ each other.

With that in mind. I think the dispatch event is the best approach.

**Comment by samant-rags** [comment]
was able to do something like this too : 
	console.log( document.querySelectorAll('.panel-content')[0].childNodes[0].childNodes[0] )

Anyways this was super helpful Zuper !
1 last Q though : Can i include external js files somewhere .. as writing everythihng in onRender() seems tedious.

I tried the template approach from the docs, but the _**_npm run build_**_ command was failing for me .. 

![image](https://user-images.githubusercontent.com/5258222/173278292-c573324b-6dc4-4e8a-9654-482a69f75539.png)

not sure why.   Tried with latest nodeJs

**Comment by ZuperZee** [comment]
Looks like the build failed, because of the lint step. (Here is a nice blog about linting https://www.freecodecamp.org/news/what-is-linting-and-how-can-it-save-you-time/).

A quick fix is to run `npm run fix`.

Another way to fix it is by removing the lint step.
Go to `package.json`
Remove `validate` from the build step in the scripts section. This `"build": "run-s validate build:prod build:custom-properties build:svg-data build:panel-options"` to this `"build": "run-s build:prod build:custom-properties build:svg-data build:panel-options"`.

Could I also know which template you tried? Looks like I need to update some of the deps on it.

**Comment by samant-rags** [comment]
silly me - i did not run the npm run fix  after upgrading nodejs to latest.  This worked !

btw for the external js  tried this approach 
In tsconfig.json added 
    ```
"allowJs": true,
    "declaration": false
```
	
In src/lib directory, created 3 files 

	addition.js	
```
		export function plus(x, y) {
			return x + y;
		}
```

	addition.d.ts
`		export declare function plus(x: number, y: number): number;
`

	index.js
`		_export * from "./addition";_
`

 npm run build ( or run :: npm run fix :: if any errors)

from the dist/panel-options.json copy content into panel import/export

**Comment by ZuperZee** [comment]
Nice. Looks like you got it working.

I made a small example https://github.com/ZuperZee/example-htmlgraphics-html-bundler-include-js/commit/e65e451bd4b4e3cb072a78276f350143d5d4b60b

_I also updated some dependencies on https://github.com/gapitio/htmlgraphics-html-bundler-template. Which might make it easier to work with :D_

Closing this as it seems resolved, but if you have any more questions it would be better to ask them here https://github.com/gapitio/gapit-htmlgraphics-panel/discussions (The discussion tab).
